mlir-sycl/doc/design/SYCLGridRepresentation.md

```diff
@@ -0,0 +1,244 @@
+| `static llvm::ArrayRef<llvm::StringLiteral> getMethodNames();` | Return the list of the method names to be replaced by this operation.  |
+| `llvm::ArrayRef<mlir::Type> getFunctionArgTypes();`            | Return the argument types of the function implementing this operation. |
+| `llvm::StringRef getFunctionName();`                           | Return the name of the function implementing this operation.           |
+| `llvm::StringRef getTypeName();`                               | Return the original name of the type this method is implemented in.    |
+
+As we can see, the attributes `FunctionName` and `TypeName` are analogous to the
+attributes present in the `sycl.call` operation, as this interface also
+represents function calls. On the other hand, the `FunctionArgTypes` attribute
+contains the argument types of the member function to call. E.g., in an
+implementation in which `nd_item` was derived from a base class and the generic
+SPIR-V address space was to be used for pointer arguments (an implementation
+different from the one corresponding to the example above), a function call to
+`nd_item::get_global_id()` would have `{!llvm.ptr<struct<...>, 4>}` as
+`FunctionArgTypes`, `@get_global_id` as `FunctionName` and `@nd_item` as
+`TypeName`.
+
+The non-member functions `getTypeID` and `getMethodNames` return respectively
+the ID of the type defined in the SYCL specification of which the function is a
+member of and a set of names which must contain `FunctionName`. Note that
+`getTypeID` does not refer to the base type present in the implementation from
+which the SYCL type is derived, but to the type present in the SYCL
+dialect. This decision removes the need of updating operations definitions if
+implementation details change or a different implementation is to be supported.
+
+### `SYCLMethodOpInterfaceImpl`
+
+To implement this interface consistently and avoid code repetition, the TableGen
+`SYCLMethodOpInterfaceImpl` class is provided. Every SYCL dialect operation
+implementing the aforementioned interface must be derived from this class.
+
+### Registering operations implementing `SYCLMethodOpInterface`
+
+The `SYCLDialect::addOperations()` function is defined, overriding
+`Dialect::addOperations()`. This function calls the base implementation, but
+also **registers** operations as `SYCLMethodOpInterface` instances. In order to
+do so, a `MethodRegistry` member variable is added to the dialect. This holds a
+`(TypeID, FunctionName) -> OperationName` table so that operations can be
+registered as "methods" calling `MethodRegistry::registerMethod(mlir::TypeID,
+llvm::StringRef, llvm::StringRef)`, where the first two parameters correspond to
+the keys of the map (`getTypeID()` and `getMethodNames()`) and the last value,
+to the value of the entry (the name of the operation).
+
+This process will require no action from the users of the dialect. As this
+mechanism is implemented in the `SYCLDialect::addOperations()` function,
```

Comment:
When is `addOperations` being invoked?

Reply:
`Dialect::addOperations()` is a function called in the initialization of every dialect. We override it with `SYCLDialect::addOperations()`.